### PR TITLE
Show grid and generator simultaneously

### PR DESCRIPTION
**Pull request by blammit**

### CODE REVIEW

File: components/widgets/WidgetHeader.qml [hunk 1]
```diff
@@ -28,13 +28,13 @@ Item {
 
 	Label {
 		id: titleLabel
-
-		anchors {
-			left: icon.right
-			leftMargin: Theme.geometry_widgetHeader_spacing
-			right: parent.right
-		}
+		width: parent.width - icon.width - secondaryLabel.implicitWidth - Theme.geometry_widgetHeader_spacing
```
Comment (jpetrell) [inline components/widgets/WidgetHeader.qml:31]: There are potentially two spacings to deduct with icon and two labels in a row?

File: pages/BriefMonitorPanel.qml [hunk 2]
```diff
@@ -23,6 +23,7 @@ Column {
 		WidgetHeader {
 			id: solarHeader
 
+			width: parent.width
```
Comment (jpetrell) [inline pages/BriefMonitorPanel.qml:26]: Are there use cases of using implicit width? If not move to the component?

File: pages/OverviewPage.qml [hunk 3]
```diff
@@ -76,6 +79,27 @@ SwipeViewPage {
...
 				break
 			case 3:
+				if (preferLargeWidgetCount === 3) {
+					// If all three prefer L size, then use M for all three.
+					widget.size = VenusOS.OverviewWidget_Size_M
+				} else if (preferLargeWidgetCount >= 2) {
+					// If only two prefer L size, then use L for the bottom one, M for the other that
+					// prefers L size, and S for the last.
+					if (i === _leftWidgets.length - 1) {
+						widget.size = VenusOS.OverviewWidget_Size_L
```
Comment (jpetrell) [inline pages/OverviewPage.qml:89]: What if the bottom one doesn't prefer the large side?

File: data/AcInputs.qml [hunk 4]
```diff
@@ -112,5 +126,13 @@ QtObject {
 		return match ? match.name : "--"
 	}
 
+	function _updateActiveInfo(info) {
+		if (info.isActiveInput) {
+			root.activeInputInfo = info
+		} else if (root.activeInputInfo == info) {
+			root.activeInputInfo = null
+		}
```
Comment (jpetrell) [inline data/AcInputs.qml:134]: Would maybe move the two signal handlers and the update function to a readonly binding:

```
readonly property AcInputSystemInfo activeInputInfo: {
    if (input1info.isActiveInput) return input1info
    else if (input2info.isActiveInput) return input2info
    else return null
}
```

File: data/common/AcInputSystemInfo.qml [hunk 5]
```diff
@@ -6,28 +6,37 @@
...
+		uid: root.bindPrefix ? root.bindPrefix + "/ServiceName" : ""
 	}
 
 	readonly property VeQuickItem _serviceType: VeQuickItem {
-		uid: root.bindPrefix + "/ServiceType"
+		uid: root.bindPrefix ? root.bindPrefix + "/ServiceType" : ""
+	}
+
+	readonly property VeQuickItem _source: VeQuickItem {
+		uid: root.bindPrefix ? root.bindPrefix + "/Source" : ""
```
Comment (jpetrell) [inline data/common/AcInputSystemInfo.qml:40]: These "root.bindPrefix ? .." checks are unfortunate, but I guess no way around them.

File: data/common/ActiveAcInput.qml [hunk 6]
```diff
@@ -9,40 +9,20 @@ import Victron.VenusOS
 Device {
 	id: root
 
-	readonly property bool connected: !!_activeInputInfo && _activeInputInfo.connected
-	readonly property string serviceType: !!_activeInputInfo ? _activeInputInfo.serviceType : ""
-	readonly property string serviceName: !!_activeInputInfo ? _activeInputInfo.serviceName : ""
+	property AcInputSystemInfo inputInfo
 
-	property int source: VenusOS.AcInputs_InputSource_NotAvailable
+	readonly property bool connected: !!inputInfo && inputInfo.connected
```
Comment (jpetrell) [inline data/common/ActiveAcInput.qml:14]: is "!!" prefix really needed?

File: data/mock/AcInputsImpl.qml [hunk 7]
```diff
@@ -11,128 +11,161 @@ QtObject {
...
+			// Cycle between positive -> negative -> zero energy.
+			// Positive energy value = imported energy, flowing towards inverter/charger.
+			// Negative energy value = exported energy, flowing towards grid.
+			const negativeEnergyFlow = testEnergyCounter < 0
+			const zeroEnergyFlow = testEnergyCounter === 0
+
+			const phases = Global.acInputs.activeInput._phases._phaseObjects
+			let totalPower = NaN
+			let i
+			for (i = 0; i < phases.count; ++i) {
```
Comment (jpetrell) [inline data/mock/AcInputsImpl.qml:140]: Why not for (let i = 0; ..

**Review by jpetrell — APPROVED** [review]
LGTM Added some small comments here and there. In testing could not find issues, works beautifully. :blush:

**Comment by jpetrell** [comment]
L1, L2 and L3 labels are shown as "L1:", "L2:" and "L3:" in the design, and without ":" here?

**Comment by jpetrell** [comment]
Why is the generator be stopped on control cards page but disconnected on the Overview page?

**Comment by blammit** [comment]
> L1, L2 and L3 labels are shown as "L1:", "L2:" and "L3:" in the design, and without ":" here?

Could you check with Serj about whether the colons are necessary? They weren't there previously and we don't use this elsewhere in the UI. Generally the color distinction is enough.

**Comment by blammit** [comment]
> Why is the generator be stopped on control cards page but disconnected on the Overview page?

Looks like it should be 'Stopped' for generators instead. Thanks, will update.

**Comment by jpetrell** [comment]
> Could you check with Serj about whether the colons are necessary? They weren't there previously and we don't use this elsewhere in the UI. Generally the color distinction is enough.

Asked Jesus, please add colons ":" to the L1-3 phase labels.

> This layout doesn't look optimal to me, though it is pretty close to the design.

Works as specified. But maybe check that "(A)" label has correct margin on the right side.

File: pages/BriefMonitorPanel.qml (hunk 2 above)
Comment (blammit) [reply pages/BriefMonitorPanel.qml]: Done

File: components/widgets/WidgetHeader.qml (hunk 1 above)
Comment (blammit) [reply components/widgets/WidgetHeader.qml]: Oh yes, thanks

File: data/AcInputs.qml (hunk 4 above)
Comment (blammit) [reply data/AcInputs.qml]: Ah yes, the function is not needed. Thanks, fixed

File: data/common/ActiveAcInput.qml (hunk 6 above)
Comment (blammit) [reply data/common/ActiveAcInput.qml]: Not here, removed

File: data/mock/AcInputsImpl.qml (hunk 7 above)
Comment (blammit) [reply data/mock/AcInputsImpl.qml]: Done

**Comment by blammit** [comment]
>  Asked Jesus, please add colons ":" to the L1-3 phase labels.

Done

> Works as specified. But maybe check that "(A)" label has correct margin on the right side.

It is the same as the left margin, but might look a bit off due to the padding within the icon image.

**Pull request merged**